### ISSUE TITLE
Importing sphinx breaks the docutils parser

### ISSUE THREAD

**Issue by brechtm** [sphinx-doc/sphinx#2799]
`test.rst`:

```
A PEP reference :PEP:`287`.
```

`test.py`:

```
from docutils.core import publish_doctree
from docutils.io import FileInput

from sphinx.application import Sphinx  # breaks docutils parsing

with open('test.rst') as file:
    doctree = publish_doctree(file, source_class=FileInput)

```

Running `test.py` fails with:

```
Traceback (most recent call last):
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/statemachine.py", line 307, in next_line
    self.line = self.input_lines[self.line_offset]
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/statemachine.py", line 1152, in __getitem__
    return self.data[i]
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/statemachine.py", line 230, in run
    self.next_line()
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/statemachine.py", line 310, in next_line
    raise EOFError
EOFError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test.py", line 8, in <module>
    doctree = publish_doctree(file, source_class=FileInput)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/core.py", line 479, in publish_doctree
    output = pub.publish(enable_exit_status=enable_exit_status)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/core.py", line 217, in publish
    self.settings)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/readers/__init__.py", line 72, in read
    self.parse()
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/readers/__init__.py", line 78, in parse
    self.parser.parse(self.input, document)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/parsers/rst/__init__.py", line 172, in parse
    self.statemachine.run(inputlines, document, inliner=self.inliner)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/parsers/rst/states.py", line 170, in run
    input_source=document['source'])
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/statemachine.py", line 245, in run
    result = state.eof(context)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/parsers/rst/states.py", line 2669, in eof
    self.blank(None, context, None)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/parsers/rst/states.py", line 2661, in blank
    context, self.state_machine.abs_line_number() - 1)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/parsers/rst/states.py", line 418, in paragraph
    textnodes, messages = self.inline_text(text, lineno)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/parsers/rst/states.py", line 427, in inline_text
    return self.inliner.parse(text, lineno, self.memo, self.parent)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/parsers/rst/states.py", line 513, in parse
    lineno)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/parsers/rst/states.py", line 776, in interpreted_or_phrase_ref
    lineno)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/docutils/parsers/rst/states.py", line 856, in interpreted
    nodes, messages2 = role_fn(role, rawsource, text, lineno, self)
  File "/Users/brechtm/.pyenv/versions/rinoh-py35/lib/python3.5/site-packages/sphinx/roles.py", line 179, in indexmarkup_role
    env = inliner.document.settings.env
AttributeError: 'Values' object has no attribute 'env'

```

**Comment by shibukawa** [comment]
Yes, it is true. Sphinx registers its own roles and directives to docutils' environment. 

https://github.com/sphinx-doc/sphinx/blob/master/sphinx/directives/__init__.py#L219-L223

And Sphinx is designed as an application, not library. If you want to use Sphinx as a library, you should initialize as same as Sphinx code:

https://github.com/sphinx-doc/sphinx/blob/master/sphinx/application.py#L103

I have never seen the document to describe how to do it.

**Comment by brechtm** [comment]
I ran into this problem in my test suite. One test calls Sphinx programmatically by instantiating the Sphinx class. Another test calls docutils' `publish_doctree`. Even when only running the latter test, pytest imports the Sphinx-based test while collecting tests.

The problem is that Sphinx overrides docutils' `pep` and `rfc` roles. I don't think Sphinx should break docutils when it is simply imported.

I don't fully understand the docutils/Sphinx internals, but a possible fix could be for `indexmarkup_role` to fall back to docutils' role-handler when the document is not a Sphinx document (I assume that's why the `AttributeError` exception is raised?).

**Comment by brechtm** [comment]
This patch makes use of the fact that Sphinx only overrides the `PEP` and `RFC` roles. These are merely aliases for `pep-reference` and `rfc-reference`. I'm not sure whether it was intended to only override the aliases, as localization is lost.

**Comment by shibukawa** [comment]
How about isolating by using multiprocessing?

```
from multiprocessing import Process, Queue
from docutils.core import publish_doctree
from docutils.io import FileInput

def sphinx_process(q):
    from sphinx.application import Sphinx
    try:
        # add any test code that uses Sphinx
        pass
    finally:
        q.put("sphinx test success")

def test_with_sphinx():
    queue = Queue()
    p = Process(target=sphinx_process, args=(queue,))
    p.start()
    result = queue.get()
    print(result)

def test_plain_docutils():
    with open('test.rst') as file:
        doctree = publish_doctree(file, source_class=FileInput)
        print("docutils test success")

def test_main():
    test_with_sphinx()
    test_plain_docutils()

if __name__ == "__main__":
    test_main()
```

**Comment by tk0miya** [comment]
Agreed. I prefer not to install roles automatically only importing sphinx module. I know sphinx is not a library. but it is a bit surprising behavior for me. So it would be nice if we improve the bootstrap process of sphinx.

On the other hand, IMO, PR #2800 is not essential resolution for this problem.
Because docutils manages the roles and directives globally. So any testcase installs docutils-extensions included in Sphinx or other extensions once, then many roles and directives will be keep installed. As a fact, Installing the extensions for docutils is strong side-effect globally. This is not resolved if we improve bootstraps of Sphinx. If you want to isolate your testcase from this side-effect, you should use multiprocessing or other techniques as @shibukawa said at this moment.

**Comment by brechtm** [comment]
The root cause of the problem is indeed the fact that registering new roles/directives affects docutils globally. I don't think this can be easily solved.

But this is really only a problem because Sphinx **overrides** the `PEP` and `RFC` roles. New roles do not interfere with the operation of docutils' parser. For this reason, I was hoping the proposed patch would be acceptable.

Thankfully, I managed to get my tests running by moving the Sphinx import into the test function. There's no need for using multiprocessing thankfully.

**Comment by tk0miya** [comment]
I fixed this at 84baf05.
It installs roles and directives on boostrap of the application.

It still overrides `pep` and `rfc` roles after application ran. I know this does not resolve completely this problem. But it might help you.
